File: packages/client-go/stream.go
package durablestreams

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"time"
)

// Protocol header names
const (
	headerContentType    = "Content-Type"
	headerStreamOffset   = "Stream-Next-Offset"
	headerStreamCursor   = "Stream-Cursor"
	headerStreamUpToDate = "Stream-Up-To-Date"
	headerStreamClosed   = "Stream-Closed"
	headerStreamSeq      = "Stream-Seq"
	headerStreamTTL      = "Stream-TTL"
	headerStreamExpires  = "Stream-Expires-At"
	headerETag           = "ETag"
	headerIfMatch        = "If-Match"
)

// Stream represents a durable stream handle.
// It is a lightweight, reusable object - not a persistent connection.
//
// Create a Stream using Client.Stream():
//
//	stream := client.Stream("https://example.com/streams/my-stream")
type Stream struct {
	url    string
	client *Client

	// Cached content type from HEAD/Create operations
	contentType string
}

// URL returns the stream's URL.
func (s *Stream) URL() string {
	return s.url
}

// ContentType returns the cached content type.
// This is populated after Create or Head operations.
func (s *Stream) ContentType() string {
	return s.contentType
}

// SetContentType sets the cached content type.
// Use this when you know the stream's content type without calling Head.
func (s *Stream) SetContentType(ct string) {
	s.contentType = ct
}

// Metadata contains stream information from HEAD request.
type Metadata struct {
	// ContentType is the stream's MIME type.
	ContentType string

	// NextOffset is the tail offset (next position after current end).
	NextOffset Offset

	// TTL is the remaining time-to-live, if set.
	TTL *time.Duration

	// ExpiresAt is the absolute expiry time, if set.
	ExpiresAt *time.Time

	// ETag for conditional requests.
	ETag string

	// StreamClosed indicates whether the stream has been closed (EOF).
	StreamClosed bool
}

// CloseResult contains the response from a close operation.
type CloseResult struct {
	// FinalOffset is the tail offset after closing the stream.
	FinalOffset Offset
}

// AppendResult contains the response from an append operation.
type AppendResult struct {
	// NextOffset is the tail offset after this append.
	// Use this for checkpointing or exactly-once semantics.
	NextOffset Offset

	// ETag for conditional requests (if returned by server).
	ETag string
}

// Create creates a new stream (idempotent).
// Succeeds if the stream already exists with matching config.
// Returns ErrStreamExists only if config differs (409 Conflict).
//
// Example:
//
//	err := stream.Create(ctx,
//	    durablestreams.WithContentType("application/json"),
//	    durablestreams.WithTTL(24*time.Hour),
//	)
func (s *Stream) Create(ctx context.Context, opts ...CreateOption) error {
	cfg := &createConfig{
		contentType: "application/octet-stream",
	}
	for _, opt := range opts {
		opt(cfg)
	}

	// Build request
	req, err := http.NewRequestWithContext(ctx, http.MethodPut, s.url, nil)
	if err != nil {
		return newStreamError("create", s.url, 0, err)
	}

	// Set headers
	req.Header.Set(headerContentType, cfg.contentType)

	if cfg.ttl > 0 {
		req.Header.Set(headerStreamTTL, strconv.FormatInt(int64(cfg.ttl.Seconds()), 10))
	}
	if !cfg.expiresAt.IsZero() {
		req.Header.Set(headerStreamExpires, cfg.expiresAt.Format(time.RFC3339))
	}
	if cfg.closed {
		req.Header.Set(headerStreamClosed, "true")
	}

	// Custom headers
	for k, v := range cfg.headers {
		req.Header.Set(k, v)
	}

	// Initial data
	if len(cfg.initialData) > 0 {
		req.Body = io.NopCloser(bytes.NewReader(cfg.initialData))
		req.ContentLength = int64(len(cfg.initialData))
	}

	// Execute request
	resp, err := s.client.httpClient.Do(req)
	if err != nil {
		return newStreamError("create", s.url, 0, err)
	}
	defer resp.Body.Close()

	// Read and discard body
	io.Copy(io.Discard, resp.Body)

	// Handle response
	switch resp.StatusCode {
	case http.StatusCreated, http.StatusOK, http.StatusNoContent:
		// Success - cache content type
		s.contentType = cfg.contentType
		return nil
	case http.StatusConflict:
		return newStreamError("create", s.url, resp.StatusCode, ErrStreamExists)
	default:
		return newStreamError("create", s.url, resp.StatusCode, errorFromStatus(resp.StatusCode))
	}
}

// Append writes data to the stream and returns the result.
// The AppendResult contains the NextOffset for checkpointing.
// Append automatically retries on transient errors (5xx, 429) with exponential backoff.
//
// Example:
//
//	result, err := stream.Append(ctx, []byte(`{"event": "test"}`))
//	fmt.Println("Next offset:", result.NextOffset)
func (s *Stream) Append(ctx context.Context, data []byte, opts ...AppendOption) (*AppendResult, error) {
	if len(data) == 0 {
		return nil, newStreamError("append", s.url, 0, ErrEmptyAppend)
	}

	cfg := &appendConfig{}
	for _, opt := range opts {
		opt(cfg)
	}

	// Set content type (use cached or default)
	contentType := s.contentType
	if contentType == "" {
		contentType = "application/octet-stream"
	}

	// Create request builder for retry
	makeRequest := func() (*http.Request, error) {
		req, err := http.NewRequestWithContext(ctx, http.MethodPost, s.url, bytes.NewReader(data))
		if err != nil {
			return nil, err
		}

		req.Header.Set(headerContentType, contentType)

		// Optional headers
		if cfg.seq != "" {
			req.Header.Set(headerStreamSeq, cfg.seq)
		}
		if cfg.ifMatch != "" {
			req.Header.Set(headerIfMatch, cfg.ifMatch)
		}

		// Custom headers
		for k, v := range cfg.headers {
			req.Header.Set(k, v)
		}

		return req, nil
	}

	// Execute with retry
	resp, err := s.doWithRetry(ctx, makeRequest)
	if err != nil {
		return nil, newStreamError("append", s.url, 0, err)
	}
	defer resp.Body.Close()

	// Read and discard body
	io.Copy(io.Discard, resp.Body)

	// Handle response
	switch resp.StatusCode {
	case http.StatusOK, http.StatusNoContent:
		return &AppendResult{
			NextOffset: Offset(resp.Header.Get(headerStreamOffset)),
			ETag:       resp.Header.Get(headerETag),
		}, nil
	case http.StatusNotFound:
		return nil, newStreamError("append", s.url, resp.StatusCode, ErrStreamNotFound)
	case http.StatusConflict:
		if resp.Header.Get(headerStreamClosed) == "true" {
			return nil, newStreamError("append", s.url, resp.StatusCode, ErrStreamClosed)
		}
		// Could be sequence conflict or content-type mismatch
		return nil, newStreamError("append", s.url, resp.StatusCode, ErrSeqConflict)
	default:
		return nil, newStreamError("append", s.url, resp.StatusCode, errorFromStatus(resp.StatusCode))
	}
}

// AppendJSON writes JSON data to the stream.
// For JSON streams, arrays are flattened one level per protocol spec.
//
// Example:
//
//	result, err := stream.AppendJSON(ctx, map[string]any{"event": "test"})
func (s *Stream) AppendJSON(ctx context.Context, v any, opts ...AppendOption) (*AppendResult, error) {
	data, err := json.Marshal(v)
	if err != nil {
		return nil, newStreamError("append", s.url, 0, fmt.Errorf("json marshal: %w", err))
	}
	return s.Append(ctx, data, opts...)
}

// Delete removes the stream.
//
// Example:
//
//	err := stream.Delete(ctx)
func (s *Stream) Delete(ctx context.Context, opts ...DeleteOption) error {
	cfg := &deleteConfig{}
	for _, opt := range opts {
		opt(cfg)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodDelete, s.url, nil)
	if err != nil {
		return newStreamError("delete", s.url, 0, err)
	}

	// Custom headers
	for k, v := range cfg.headers {
		req.Header.Set(k, v)
	}

	resp, err := s.client.httpClient.Do(req)
	if err != nil {
		return newStreamError("delete", s.url, 0, err)
	}
	defer resp.Body.Close()

	io.Copy(io.Discard, resp.Body)

	switch resp.StatusCode {
	case http.StatusOK, http.StatusNoContent:
		return nil
	case http.StatusNotFound:
		return newStreamError("delete", s.url, resp.StatusCode, ErrStreamNotFound)
	default:
		return newStreamError("delete", s.url, resp.StatusCode, errorFromStatus(resp.StatusCode))
	}
}

// Close closes the stream, optionally with a final message.
//
// After closing:
//   - No further appends are permitted (server returns 409)
//   - Readers can observe the closed state and treat it as EOF
//   - The stream's data remains fully readable
//
// Closing is:
//   - Durable: The closed state is persisted
//   - Monotonic: Once closed, a stream cannot be reopened
//   - Idempotent (without body): Safe to call multiple times
//
// Example:
//
//	result, err := stream.Close(ctx)
//	result, err := stream.Close(ctx, WithCloseData([]byte("final")))
func (s *Stream) Close(ctx context.Context, opts ...CloseOption) (*CloseResult, error) {
	cfg := &closeConfig{}
	for _, opt := range opts {
		opt(cfg)
	}

	var body io.Reader
	var contentLength int64
	if len(cfg.data) > 0 {
		// For JSON streams, wrap in array
		contentType := cfg.contentType
		if contentType == "" {
			contentType = s.contentType
		}
		if contentType == "application/json" {
			wrapped := []byte("[")
			wrapped = append(wrapped, cfg.data...)
			wrapped = append(wrapped, ']')
			body = bytes.NewReader(wrapped)
			contentLength = int64(len(wrapped))
		} else {
			body = bytes.NewReader(cfg.data)
			contentLength = int64(len(cfg.data))
		}
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, s.url, body)
	if err != nil {
		return nil, newStreamError("close", s.url, 0, err)
	}

	// Set headers
	req.Header.Set(headerStreamClosed, "true")
	if contentLength > 0 {
		req.ContentLength = contentLength
	}

	contentType := cfg.contentType
	if contentType == "" {
		contentType = s.contentType
	}
	if contentType != "" {
		req.Header.Set(headerContentType, contentType)
	}

	// Custom headers
	for k, v := range cfg.headers {
		req.Header.Set(k, v)
	}

	resp, err := s.client.httpClient.Do(req)
	if err != nil {
		return nil, newStreamError("close", s.url, 0, err)
	}
	defer resp.Body.Close()

	io.Copy(io.Discard, resp.Body)

	// Check for 409 Conflict with Stream-Closed header (stream was already closed)
	if resp.StatusCode == http.StatusConflict {
		isClosed := resp.Header.Get(headerStreamClosed) == "true"
		if isClosed {
			return nil, newStreamError("close", s.url, resp.StatusCode, ErrStreamClosed)
		}
		return nil, newStreamError("close", s.url, resp.StatusCode, ErrSeqConflict)
	}

	switch resp.StatusCode {
	case http.StatusOK, http.StatusNoContent:
		return &CloseResult{
			FinalOffset: Offset(resp.Header.Get(headerStreamOffset)),
		}, nil
	case http.StatusNotFound:
		return nil, newStreamError("close", s.url, resp.StatusCode, ErrStreamNotFound)
	default:
		return nil, newStreamError("close", s.url, resp.StatusCode, errorFromStatus(resp.StatusCode))
	}
}

// Head returns stream metadata without reading content.
//
// Example:
//
//	meta, err := stream.Head(ctx)
//	fmt.Println("Content-Type:", meta.ContentType)
//	fmt.Println("Next offset:", meta.NextOffset)
func (s *Stream) Head(ctx context.Context, opts ...HeadOption) (*Metadata, error) {
	cfg := &headConfig{}
	for _, opt := range opts {
		opt(cfg)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodHead, s.url, nil)
	if err != nil {
		return nil, newStreamError("head", s.url, 0, err)
	}

	// Custom headers
	for k, v := range cfg.headers {
		req.Header.Set(k, v)
	}

	resp, err := s.client.httpClient.Do(req)
	if err != nil {
		return nil, newStreamError("head", s.url, 0, err)
	}
	defer resp.Body.Close()

	switch resp.StatusCode {
	case http.StatusOK:
		meta := &Metadata{
			ContentType:  resp.Header.Get(headerContentType),
			NextOffset:   Offset(resp.Header.Get(headerStreamOffset)),
			ETag:         resp.Header.Get(headerETag),
			StreamClosed: resp.Header.Get(headerStreamClosed) == "true",
		}

		// Cache content type
		if meta.ContentType != "" {
			s.contentType = meta.ContentType
		}

		// Parse optional TTL
		if ttlStr := resp.Header.Get(headerStreamTTL); ttlStr != "" {
			if secs, err := strconv.ParseInt(ttlStr, 10, 64); err == nil {
				ttl := time.Duration(secs) * time.Second
				meta.TTL = &ttl
			}
		}

		// Parse optional expiry
		if expiresStr := resp.Header.Get(headerStreamExpires); expiresStr != "" {
			if t, err := time.Parse(time.RFC3339, expiresStr); err == nil {
				meta.ExpiresAt = &t
			}
		}

		return meta, nil
	case http.StatusNotFound:
		return nil, newStreamError("head", s.url, resp.StatusCode, ErrStreamNotFound)
	default:
		return nil, newStreamError("head", s.url, resp.StatusCode, errorFromStatus(resp.StatusCode))
	}
}

// Read returns an iterator for reading stream chunks.
// Each chunk corresponds to one HTTP response body.
// The iterator handles catch-up, live tailing, and cursor propagation automatically.
//
// Always call Close() when done:
//
//	it := stream.Read(ctx)
//	defer it.Close()
//
//	for {
//	    chunk, err := it.Next()
//	    if errors.Is(err, durablestreams.Done) {
//	        break
//	    }
//	    if err != nil {
//	        return err
//	    }
//	    // Process chunk.Data
//	}
func (s *Stream) Read(ctx context.Context, opts ...ReadOption) *ChunkIterator {
	cfg := &readConfig{
		offset:  StartOffset,
		live:    LiveModeNone,
		timeout: 30 * time.Second,
	}
	for _, opt := range opts {
		opt(cfg)
	}

	// Create a cancellable context for the iterator
	iterCtx, cancel := context.WithCancel(ctx)

	return &ChunkIterator{
		stream:  s,
		ctx:     iterCtx,
		cancel:  cancel,
		offset:  cfg.offset,
		live:    cfg.live,
		cursor:  cfg.cursor,
		headers: cfg.headers,
		timeout: cfg.timeout,
		Offset:  cfg.offset,
		UpToDate: false,
	}
}

// buildReadURL constructs the URL for a read request with query parameters.
func (s *Stream) buildReadURL(offset Offset, live LiveMode, cursor string) string {
	u, err := url.Parse(s.url)
	if err != nil {
		return s.url
	}

	q := u.Query()

	// Always include offset (even for start position "-1")
	if offset.IsStart() {
		q.Set("offset", string(StartOffset))
	} else {
		q.Set("offset", string(offset))
	}

	// Add live mode
	switch live {
	case LiveModeLongPoll:
		q.Set("live", "long-poll")
	case LiveModeSSE:
		q.Set("live", "sse")
	}

	// Add cursor for CDN collapsing
	if cursor != "" {
		q.Set("cursor", cursor)
	}

	u.RawQuery = q.Encode()
	return u.String()
}

File: packages/client-rust/src/stream.rs
//! Stream handle and operations.

use crate::client::Client;
use crate::error::StreamError;
use crate::iterator::ReadBuilder;
use crate::producer::ProducerBuilder;
use crate::types::Offset;
use bytes::Bytes;
use std::time::Duration;

/// Protocol header names
pub(crate) const HEADER_CONTENT_TYPE: &str = "content-type";
pub(crate) const HEADER_STREAM_OFFSET: &str = "stream-next-offset";
pub(crate) const HEADER_STREAM_CURSOR: &str = "stream-cursor";
pub(crate) const HEADER_STREAM_UP_TO_DATE: &str = "stream-up-to-date";
pub(crate) const HEADER_STREAM_SEQ: &str = "stream-seq";
pub(crate) const HEADER_STREAM_TTL: &str = "stream-ttl";
pub(crate) const HEADER_STREAM_EXPIRES: &str = "stream-expires-at";
pub(crate) const HEADER_ETAG: &str = "etag";
pub(crate) const HEADER_IF_MATCH: &str = "if-match";

/// Producer headers
pub(crate) const HEADER_PRODUCER_ID: &str = "producer-id";
pub(crate) const HEADER_PRODUCER_EPOCH: &str = "producer-epoch";
pub(crate) const HEADER_PRODUCER_SEQ: &str = "producer-seq";
pub(crate) const HEADER_PRODUCER_EXPECTED_SEQ: &str = "producer-expected-seq";
pub(crate) const HEADER_STREAM_CLOSED: &str = "stream-closed";

/// Maximum retries for transient errors on append operations
const MAX_APPEND_RETRIES: u32 = 3;

/// A handle to a durable stream.
///
/// This is a lightweight, cloneable object - not a persistent connection.
/// Operations make HTTP requests on demand.
#[derive(Clone, Debug)]
pub struct DurableStream {
    pub(crate) url: String,
    pub(crate) client: Client,
    pub(crate) content_type: Option<String>,
}

impl DurableStream {
    /// Get the stream URL.
    pub fn url(&self) -> &str {
        &self.url
    }

    /// Get the content type set on this stream handle.
    ///
    /// This is used as the default Content-Type for append operations
    /// and by the Producer for JSON mode detection.
    ///
    /// Note: This is not automatically populated from the server.
    /// Use [`set_content_type`](Self::set_content_type) to set it after
    /// creating a stream, or set it explicitly on the Producer.
    pub fn content_type(&self) -> Option<&str> {
        self.content_type.as_deref()
    }

    /// Set the content type for this stream handle.
    ///
    /// This affects append operations and Producer JSON mode detection.
    ///
    /// # Example
    /// ```ignore
    /// let mut stream = client.stream("...");
    /// stream.set_content_type("application/json");
    /// ```
    pub fn set_content_type(&mut self, ct: impl Into<String>) {
        self.content_type = Some(ct.into());
    }

    /// Create the stream.
    ///
    /// Idempotent - succeeds if stream already exists with matching config.
    /// Returns `StreamError::Conflict` only if config differs.
    pub async fn create(&self) -> Result<(), StreamError> {
        self.create_with(CreateOptions::default()).await
    }

    /// Create the stream with options.
    pub async fn create_with(&self, options: CreateOptions) -> Result<(), StreamError> {
        let content_type = options
            .content_type
            .as_deref()
            .unwrap_or("application/octet-stream");

        let mut req = self
            .client
            .inner
            .put(&self.url)
            .header(HEADER_CONTENT_TYPE, content_type);

        // Add TTL header if specified
        if let Some(ttl) = options.ttl {
            req = req.header(HEADER_STREAM_TTL, ttl.as_secs().to_string());
        }

        // Add expires header if specified
        if let Some(expires) = &options.expires_at {
            req = req.header(HEADER_STREAM_EXPIRES, expires);
        }

        // Add custom headers
        let client_headers = self.client.get_headers();
        for (key, value) in client_headers.iter() {
            req = req.header(key.clone(), value.clone());
        }

        for (key, value) in &options.headers {
            req = req.header(key.as_str(), value.as_str());
        }

        // Add closed header if specified
        if options.closed {
            req = req.header(HEADER_STREAM_CLOSED, "true");
        }

        // Add initial data if provided
        if let Some(data) = options.initial_data {
            req = req.body(data);
        }

        let resp = req.send().await?;
        let status = resp.status().as_u16();

        match status {
            200 | 201 | 204 => Ok(()),
            409 => Err(StreamError::Conflict),
            _ => Err(StreamError::from_status(status, &self.url)),
        }
    }

    /// Append data to the stream.
    pub async fn append(&self, data: impl Into<Bytes>) -> Result<AppendResponse, StreamError> {
        self.append_with(data, AppendOptions::default()).await
    }

    /// Append data with options.
    pub async fn append_with(
        &self,
        data: impl Into<Bytes>,
        options: AppendOptions,
    ) -> Result<AppendResponse, StreamError> {
        let data = data.into();
        if data.is_empty() {
            return Err(StreamError::EmptyAppend);
        }

        let content_type = self
            .content_type
            .as_deref()
            .unwrap_or("application/octet-stream");

        // Retry logic for transient errors
        let mut last_error = None;

        for attempt in 0..=MAX_APPEND_RETRIES {
            if attempt > 0 {
                // Exponential backoff: 100ms, 200ms, 400ms
                tokio::time::sleep(std::time::Duration::from_millis(100 * (1 << (attempt - 1)))).await;
            }

            let mut req = self
                .client
                .inner
                .post(&self.url)
                .header(HEADER_CONTENT_TYPE, content_type)
                .body(data.clone());

            // Add sequence header if specified
            if let Some(seq) = &options.seq {
                req = req.header(HEADER_STREAM_SEQ, seq.as_str());
            }

            // Add if-match header if specified
            if let Some(etag) = &options.if_match {
                req = req.header(HEADER_IF_MATCH, etag.as_str());
            }

            // Add custom headers
            let client_headers = self.client.get_headers();
            for (key, value) in client_headers.iter() {
                req = req.header(key.clone(), value.clone());
            }

            for (key, value) in &options.headers {
                req = req.header(key.as_str(), value.as_str());
            }

            let resp = match req.send().await {
                Ok(r) => r,
                Err(e) => {
                    last_error = Some(StreamError::from(e));
                    continue; // Retry on network error
                }
            };

            let status = resp.status().as_u16();

            match status {
                200 | 204 => {
                    let next_offset = resp
                        .headers()
                        .get(HEADER_STREAM_OFFSET)
                        .and_then(|v| v.to_str().ok())
                        .map(|s| Offset::parse(s))
                        .unwrap_or(Offset::Beginning);

                    let etag = resp
                        .headers()
                        .get(HEADER_ETAG)
                        .and_then(|v| v.to_str().ok())
                        .map(|s| s.to_string());

                    return Ok(AppendResponse { next_offset, etag });
                }
                404 => return Err(StreamError::NotFound {
                    url: self.url.clone(),
                }),
                409 => {
                    let stream_closed = resp
                        .headers()
                        .get(HEADER_STREAM_CLOSED)
                        .and_then(|v| v.to_str().ok())
                        .map(|s| s.eq_ignore_ascii_case("true"))
                        .unwrap_or(false);
                    if stream_closed {
                        return Err(StreamError::StreamClosed);
                    }
                    return Err(StreamError::SeqConflict);
                }
                // Retry on transient server errors
                500 | 502 | 503 | 504 | 429 => {
                    last_error = Some(StreamError::from_status(status, &self.url));
                    continue;
                }
                _ => return Err(StreamError::from_status(status, &self.url)),
            }
        }

        // Return last error if all retries failed
        Err(last_error.unwrap_or_else(|| StreamError::ServerError {
            status: 500,
            message: "All retries failed".to_string(),
        }))
    }

    /// Get stream metadata via HEAD request.
    pub async fn head(&self) -> Result<HeadResponse, StreamError> {
        self.head_with(HeadOptions::default()).await
    }

    /// Get stream metadata with options.
    pub async fn head_with(&self, options: HeadOptions) -> Result<HeadResponse, StreamError> {
        let mut req = self.client.inner.head(&self.url);

        // Add custom headers
        let client_headers = self.client.get_headers();
        for (key, value) in client_headers.iter() {
            req = req.header(key.clone(), value.clone());
        }

        for (key, value) in &options.headers {
            req = req.header(key.as_str(), value.as_str());
        }

        let resp = req.send().await?;
        let status = resp.status().as_u16();

        match status {
            200 => {
                let next_offset = resp
                    .headers()
                    .get(HEADER_STREAM_OFFSET)
                    .and_then(|v| v.to_str().ok())
                    .map(|s| Offset::parse(s))
                    .unwrap_or(Offset::Beginning);

                let content_type = resp
                    .headers()
                    .get(HEADER_CONTENT_TYPE)
                    .and_then(|v| v.to_str().ok())
                    .map(|s| s.to_string());

                let ttl = resp
                    .headers()
                    .get(HEADER_STREAM_TTL)
                    .and_then(|v| v.to_str().ok())
                    .and_then(|s| s.parse::<u64>().ok())
                    .map(Duration::from_secs);

                let expires_at = resp
                    .headers()
                    .get(HEADER_STREAM_EXPIRES)
                    .and_then(|v| v.to_str().ok())
                    .map(|s| s.to_string());

                let etag = resp
                    .headers()
                    .get(HEADER_ETAG)
                    .and_then(|v| v.to_str().ok())
                    .map(|s| s.to_string());

                let stream_closed = resp
                    .headers()
                    .get(HEADER_STREAM_CLOSED)
                    .and_then(|v| v.to_str().ok())
                    .map(|s| s.eq_ignore_ascii_case("true"))
                    .unwrap_or(false);

                Ok(HeadResponse {
                    next_offset,
                    content_type,
                    ttl,
                    expires_at,
                    etag,
                    stream_closed,
                })
            }
            404 => Err(StreamError::NotFound {
                url: self.url.clone(),
            }),
            _ => Err(StreamError::from_status(status, &self.url)),
        }
    }

    /// Delete the stream.
    pub async fn delete(&self) -> Result<(), StreamError> {
        self.delete_with(DeleteOptions::default()).await
    }

    /// Delete the stream with options.
    pub async fn delete_with(&self, options: DeleteOptions) -> Result<(), StreamError> {
        let mut req = self.client.inner.delete(&self.url);

        // Add custom headers
        let client_headers = self.client.get_headers();
        for (key, value) in client_headers.iter() {
            req = req.header(key.clone(), value.clone());
        }

        for (key, value) in &options.headers {
            req = req.header(key.as_str(), value.as_str());
        }

        let resp = req.send().await?;
        let status = resp.status().as_u16();

        match status {
            200 | 204 => Ok(()),
            404 => Err(StreamError::NotFound {
                url: self.url.clone(),
            }),
            _ => Err(StreamError::from_status(status, &self.url)),
        }
    }

    /// Close the stream (no more appends allowed).
    pub async fn close(&self) -> Result<CloseResponse, StreamError> {
        self.close_with(CloseOptions::default()).await
    }

    /// Close the stream with options.
    pub async fn close_with(&self, options: CloseOptions) -> Result<CloseResponse, StreamError> {
        let content_type = options
            .content_type
            .as_deref()
            .or(self.content_type.as_deref())
            .unwrap_or("application/octet-stream");

        let mut req = self.client.inner.post(&self.url);

        // Add custom headers
        let client_headers = self.client.get_headers();
        for (key, value) in client_headers.iter() {
            req = req.header(key.clone(), value.clone());
        }

        for (key, value) in &options.headers {
            req = req.header(key.as_str(), value.as_str());
        }

        req = req.header(HEADER_STREAM_CLOSED, "true");
        req = req.header(HEADER_CONTENT_TYPE, content_type);

        // Add data if provided
        if let Some(data) = options.data {
            // For JSON streams, wrap data in array
            let body = if content_type.to_lowercase().contains("application/json") {
                let mut wrapped = Vec::with_capacity(data.len() + 2);
                wrapped.push(b'[');
                wrapped.extend_from_slice(&data);
                wrapped.push(b']');
                Bytes::from(wrapped)
            } else {
                data
            };
            req = req.body(body);
        }

        let resp = req.send().await?;
        let status = resp.status().as_u16();

        match status {
            200 | 204 => {
                let final_offset = resp
                    .headers()
                    .get(HEADER_STREAM_OFFSET)
                    .and_then(|v| v.to_str().ok())
                    .map(|s| Offset::parse(s))
                    .unwrap_or(Offset::Beginning);

                Ok(CloseResponse { final_offset })
            }
            404 => Err(StreamError::NotFound {
                url: self.url.clone(),
            }),
            409 => {
                let stream_closed = resp
                    .headers()
                    .get(HEADER_STREAM_CLOSED)
                    .and_then(|v| v.to_str().ok())
                    .map(|s| s.eq_ignore_ascii_case("true"))
                    .unwrap_or(false);

                if stream_closed {
                    Err(StreamError::StreamClosed)
                } else {
                    Err(StreamError::SeqConflict)
                }
            }
            _ => Err(StreamError::from_status(status, &self.url)),
        }
    }

    /// Create a reader builder for consuming the stream.
    pub fn read(&self) -> ReadBuilder {
        ReadBuilder::new(self.clone())
    }

    /// Create an idempotent producer builder.
    pub fn producer(&self, producer_id: impl Into<String>) -> ProducerBuilder {
        ProducerBuilder::new(self.clone(), producer_id.into())
    }

    /// Build a read URL with query parameters.
    pub(crate) fn build_read_url(
        &self,
        offset: &Offset,
        live: Option<&str>,
        cursor: Option<&str>,
    ) -> String {
        let mut url = self.url.clone();
        let mut params = Vec::new();

        // Always include offset
        params.push(format!("offset={}", offset.to_query_value()));

        // Add live mode if specified
        if let Some(live) = live {
            params.push(format!("live={}", live));
        }

        // Add cursor if specified
        if let Some(cursor) = cursor {
            params.push(format!("cursor={}", cursor));
        }

        if !params.is_empty() {
            if url.contains('?') {
                url.push('&');
            } else {
                url.push('?');
            }
            url.push_str(&params.join("&"));
        }

        url
    }
}

/// Options for creating a stream.
#[derive(Clone, Debug, Default)]
#[non_exhaustive]
pub struct CreateOptions {
    pub content_type: Option<String>,
    pub ttl: Option<Duration>,
    pub expires_at: Option<String>,
    pub headers: Vec<(String, String)>,
    pub initial_data: Option<Bytes>,
    pub closed: bool,
}

impl CreateOptions {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn content_type(mut self, ct: impl Into<String>) -> Self {
        self.content_type = Some(ct.into());
        self
    }

    pub fn ttl(mut self, ttl: Duration) -> Self {
        self.ttl = Some(ttl);
        self
    }

    pub fn expires_at(mut self, expires: impl Into<String>) -> Self {
        self.expires_at = Some(expires.into());
        self
    }

    pub fn header(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.headers.push((key.into(), value.into()));
        self
    }

    pub fn initial_data(mut self, data: impl Into<Bytes>) -> Self {
        self.initial_data = Some(data.into());
        self
    }

    pub fn closed(mut self, closed: bool) -> Self {
        self.closed = closed;
        self
    }
}

/// Options for appending to a stream.
#[derive(Clone, Debug, Default)]
#[non_exhaustive]
pub struct AppendOptions {
    pub seq: Option<String>,
    pub if_match: Option<String>,
    pub headers: Vec<(String, String)>,
}

impl AppendOptions {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn seq(mut self, seq: impl Into<String>) -> Self {
        self.seq = Some(seq.into());
        self
    }

    pub fn if_match(mut self, etag: impl Into<String>) -> Self {
        self.if_match = Some(etag.into());
        self
    }

    pub fn header(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.headers.push((key.into(), value.into()));
        self
    }
}

/// Options for HEAD request.
#[derive(Clone, Debug, Default)]
pub struct HeadOptions {
    pub headers: Vec<(String, String)>,
}

impl HeadOptions {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn header(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.headers.push((key.into(), value.into()));
        self
    }
}

/// Options for DELETE request.
#[derive(Clone, Debug, Default)]
pub struct DeleteOptions {
    pub headers: Vec<(String, String)>,
}

impl DeleteOptions {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn header(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.headers.push((key.into(), value.into()));
        self
    }
}

/// Response from an append operation.
#[derive(Clone, Debug)]
#[non_exhaustive]
pub struct AppendResponse {
    pub next_offset: Offset,
    pub etag: Option<String>,
}

/// Response from a HEAD operation.
#[derive(Clone, Debug)]
#[non_exhaustive]
pub struct HeadResponse {
    pub next_offset: Offset,
    pub content_type: Option<String>,
    pub ttl: Option<Duration>,
    pub expires_at: Option<String>,
    pub etag: Option<String>,
    pub stream_closed: bool,
}

/// Response from a close operation.
#[derive(Clone, Debug)]
#[non_exhaustive]
pub struct CloseResponse {
    pub final_offset: Offset,
}

/// Options for closing a stream.
#[derive(Clone, Debug, Default)]
#[non_exhaustive]
pub struct CloseOptions {
    pub data: Option<Bytes>,
    pub content_type: Option<String>,
    pub headers: Vec<(String, String)>,
}

impl CloseOptions {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn data(mut self, data: impl Into<Bytes>) -> Self {
        self.data = Some(data.into());
        self
    }

    pub fn content_type(mut self, ct: impl Into<String>) -> Self {
        self.content_type = Some(ct.into());
        self
    }

    pub fn header(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.headers.push((key.into(), value.into()));
        self
    }
}
